SPLAT-1173: Enhance vSphere Installer to use IPAddressClaims for static IP

vr4manta opened the pull request:
[SPLAT-1173](https://issues.redhat.com//browse/SPLAT-1173)

## Changes
- Enhance installer to generate IPAddress and IPAddressClaim for each IP being associated to device 0
- Enhance terraform logic to leverage the IPAddress object as the source-of-truth for generating the kargs for static IP of each master machine.

## Dependencies
- https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/273

## Notes
- The fix in the [CPMSO](https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/273) is required due to CPMSO not knowing how to handle AddressesFromPools in its comparisons.

rvanderp3 commented on pkg/asset/machines/master.go:
```diff
@@ -583,6 +596,33 @@ func (m *Master) Generate(dependencies asset.Parents) error {
+	for i, address := range ipAddrs {
+		data, err := yaml.Marshal(address)
+		if err != nil {
+			return errors.Wrapf(err, "marshal ip claim %v", address.Name)
```
unable to marshal ipaddress %v

rvanderp3 commented on pkg/asset/machines/master.go:
```diff
@@ -583,6 +596,33 @@ func (m *Master) Generate(dependencies asset.Parents) error {
+	for i, claim := range ipClaims {
+		data, err := yaml.Marshal(claim)
+		if err != nil {
+			return errors.Wrapf(err, "marshal ip claim %v", claim.Name)
```
unable to marshal ipaddressclaim %v

rvanderp3 commented on pkg/asset/machines/master.go:
```diff
@@ -752,13 +808,40 @@ func IsMachineManifest(file *asset.File) bool {
+		address := &ipamv1.IPAddress{}
+		err := yaml.Unmarshal(file.Data, &address)
+		if err != nil {
+			return ipAddresses, errors.Wrapf(err, "unmarshal ip address %d", i)
```
change to: "unable to unmarshal ipaddress %d"

vr4manta replied:
I was using same error message from the lines prior to this where the same activities are being done.  For consistency, should I also update the other error messages to add "unable to "?

rvanderp3 replied:
i dont think that's necessary, was just thinking it might clarify the error.

rvanderp3 reviewed (commented):
/test lint

rvanderp3 commented:
/test go-lint

rvanderp3 commented:
/test golint

vr4manta commented:
/hold

vr4manta commented:
/hold cancel

rvanderp3 commented:
/hold

pending #7418

vr4manta commented:
/reopen

patrickdillon commented on pkg/tfvars/vsphere/vsphere_test.go:
```diff
@@ -83,24 +93,65 @@ func createControlPlaneConfigs(ipTypes int64) []*v1beta1.VSphereMachineProviderS
-			machine.Network.Devices[0].IPAddrs = append(machine.Network.Devices[0].IPAddrs, fmt.Sprintf(ipv4Template, i))
+			ipAddress := ipamv1.IPAddress{}
+			ipAddress.Name = fmt.Sprintf("%s-claim-%d-%d", machine.Name, 0, idx)
+			pool := v1beta1.AddressesFromPool{
+				Group:    "installer.openshift.io",
+				Resource: "IPPool",
+				Name:     fmt.Sprintf("default-%d", idx),
+			}
```
This `installer.openshift.io` pool would not exist (by design); at a minimum we will document this, but need to consider/review a bit further for possible side effects.

vr4manta replied:
So far I have not seen anything negative.  We do have a card to follow up on this.

patrickdillon commented:
/assign

vr4manta commented:
/assign @patrickdillon

vr4manta commented:
/payload-job-with-prs periodic-ci-openshift-release-master-nightly-4.16-e2e-vsphere-static-ovn

vr4manta commented:
/hold cancel

vr4manta commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.16-e2e-vsphere-static-ovn

vr4manta commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.16-e2e-vsphere-static-ovn

vr4manta commented:
/retest-required

vr4manta commented:
@JoelSpeed @patrickdillon , I'd like to have this PR reviewed now that we've fixed static IP periodicals and fixed issues related to static IP with CPMS.

rvanderp3 commented on pkg/asset/machines/vsphere/capimachines.go:
```diff
@@ -57,7 +61,7 @@ func getNetworkInventoryPath(vcenterContext vsphere.VCenterContext, networkName
 // GenerateMachines returns a list of capi machines.
 func GenerateMachines(ctx context.Context, clusterID string, config *types.InstallConfig, pool *types.MachinePool, osImage string, role string, metadata *vsphere.Metadata) ([]*asset.RuntimeFile, error) {
-	machines, _, err := Machines(clusterID, config, pool, osImage, role, "")
+	machines, _, _, _, err := Machines(clusterID, config, pool, osImage, role, "") //nolint:dogsled
```
can the code be refactored such that this nolint isnt needed or is there a firm reason to keep it?

vr4manta replied:
If it gets refactored, do as separate PR.  The CAPI env was not what the Machine interface was intended for.  I am not sure why it was reused instead of consuming what may have been generated in the normal machine creation flow.

vr4manta replied:
This has now been refactored and nolint has been removed.

rvanderp3 commented:
/hold
/lgtm
/approve
overall lgtm, i added one comment.  depending on what you think feel free to drop the hold.

rvanderp3 commented:
/lgtm

jcpowermac commented:
/test altinfra-e2e-vsphere-capi-ovn

vr4manta commented:
/retest

vr4manta commented:
/retest

jcpowermac commented:
/lgtm

vr4manta commented:
/retest

jcpowermac commented:
/lgtm

patrickdillon commented:
/approve

vr4manta commented:
/retest-required

vr4manta commented:
looks like I'm hitting the issue fixed by https://github.com/openshift/origin/pull/28631

r4f4 commented:
/override ci/prow/e2e-aws-ovn ci/prow/e2e-vsphere-ovn
